librad/src/meta/entity.rs

```diff
@@ -145,31 +206,113 @@ pub struct Signed;
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct Draft;
 
-/// A type expressing *who* is signing an `Entity`
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub enum Signatory {
-    /// A specific user (identified by their URN)
-    User(RadUrn),
-    /// The entity itself (with an owned key)
-    OwnedKey,
+fn build_signature_data(
+    hash: &Hash,
+    revision: EntityRevision,
+    timestamp: EntityTimestamp,
+) -> Vec<u8> {
+    let mut data: Vec<u8> = Vec::with_capacity(hash.as_bytes().len() + 8 + 8);
+    data.extend_from_slice(&revision.to_ne_bytes());
+    data.extend_from_slice(&timestamp.epoch_millis().to_ne_bytes());
+    data.extend_from_slice(hash.as_bytes());
+    data
 }
 
-/// A signature for an `Entity``
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub struct EntitySignature {
-    /// Who is producing this signature
-    pub by: Signatory,
+fn build_signature(
+    key: &SecretKey,
+    hash: &Hash,
+    revision: EntityRevision,
+    timestamp: EntityTimestamp,
+) -> Signature {
+    let data = build_signature_data(hash, revision, timestamp);
+    key.sign(&data)
+}
+
+fn verify_signature(
+    sig: &Signature,
+    key: &PublicKey,
+    hash: &Hash,
+    revision: EntityRevision,
+    timestamp: EntityTimestamp,
+) -> bool {
+    let data = build_signature_data(hash, revision, timestamp);
+    sig.verify(&data, key)
+}
+
+/// A signature for an `Entity`, when signed using an owned key
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SelfSignature {
+    /// Signature time
+    pub timestamp: EntityTimestamp,
     /// The signature data
     pub sig: Signature,
 }
 
+impl SelfSignature {
+    fn build(
+        key: &SecretKey,
+        hash: &Hash,
+        revision: EntityRevision,
+        timestamp: EntityTimestamp,
+    ) -> Self {
+        let sig = build_signature(key, hash, revision, timestamp);
+        Self { timestamp, sig }
+    }
+
+    fn new(key: &SecretKey, hash: &Hash, revision: EntityRevision) -> Self {
+        Self::build(key, hash, revision, EntityTimestamp::current_time())
+    }
+
+    pub(crate) fn verify(&self, key: &PublicKey, hash: &Hash, revision: EntityRevision) -> bool {
+        verify_signature(&self.sig, key, hash, revision, self.timestamp)
+    }
+}
+
+/// A signature for an `Entity`, when signed by a certifier
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct CertifierSignature {
```

Comment:
I'm not understanding the difference between `SelfSignature` and `CertifierSignature`. They look like they're built from the same data but `SelfSignature` doesn't hold onto the details. Why is there even a difference?